### PR TITLE
Repair some corner cases in cancellation propagation between coroutines and listenable futures

### PR DESCRIPTION
**Pull request by vadimsemenov**
Implement bidirectional cancellation for future coroutine builder.

This also:
- Refactors JobListenableFuture infrastructure so it can be reused in CoroutineScope.future and Deferred.asListenableFuture;
- Hides ListenableFuture.asDeferred return type, so it can't be casted to CompletableDeferred;
- Adds more tests to cover fixed corner cases;
- Improves documentation;
- Suppresses annoying warnings in tests.

### CODE REVIEW

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 1]
```diff
@@ -33,7 +35,10 @@ import kotlin.coroutines.*
  * facilities.
  *
  * Note that the error and cancellation semantics of [future] are _subtly different_ than
- * [asListenableFuture]'s. See [ListenableFutureCoroutine] for details.
+ * [asListenableFuture]'s.
```
Comment (yorickhenning) [inline integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: Remove the linebreak here.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 2]
```diff
@@ -17,8 +17,10 @@ import kotlin.coroutines.*
  * The coroutine is immediately started. Passing [CoroutineStart.LAZY] to [start] throws
  * [IllegalArgumentException], because Futures don't have a way to start lazily.
  *
- * The created coroutine is cancelled when the resulting future completes successfully, fails, or
- * is cancelled.
+ * When the created coroutine completes (either successfully or exceptionally), it will try to
```
Comment (yorickhenning) [inline integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: Avoid parentheticals in docstrings, and in technical writing generally. Instead, use several shorter sentences and restate the direct object as necessary.

"When the created coroutine finishes, it *synchronously* completes the returned future with the same outcome. The completion is synchronous whether the coroutine succeeds or fails."

Comment (yorickhenning) [inline integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: Is there any way completing the Future will fail? The LF is settable/completable, so this should be able to categorically state that completion is synchronous.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 3]
```diff
@@ -421,42 +417,64 @@ private class OuterFuture<T>(private val deferred: Deferred<T>): ListenableFutur
...
+    @Suppress("UnstableApiUsage")
+    override fun afterDone() {
+        if (isCancelled) {
+            jobToCancel.cancel()
         }
     }
 }
+
+/** Represents result of a `Coroutine` (either [Deferred] or [AbstractCoroutine]). */
+// This is similar to kotlin.Result type. However, kotlin.Result can't be used by external libraries.
```
Comment (yorickhenning) [inline integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: It can't? It's an inline class in the core SDK. It's not a JVM-specific type.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 4]
```diff
@@ -394,17 +398,9 @@ private class OuterFuture<T>(private val deferred: Deferred<T>): ListenableFutur
...
-                throw exception
-            } else {
-                throw ExecutionException(exception)
-            }
-        } else {
-            return deferred.getCompleted()
-        }
+    private fun getInternal(result: Result<T>): T = when(result) {
+        is Result.Success -> result.value
+        is Result.Failure -> throw result.asFutureException()
```
Comment (yorickhenning) [inline integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: This reallocates an Exception each time the Future is observed. Memoizing the exception is probably a good idea.

I hope no one cares about the reference equality of the Exceptions thrown, but it's not good to generate multiple distinct exception objects from a single failure.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 5]
```diff
@@ -421,42 +417,64 @@ private class OuterFuture<T>(private val deferred: Deferred<T>): ListenableFutur
...
      * that certain combinations of cancelled/cancelling states can't be observed.
      */
-    override fun cancel(mayInterruptIfRunning: Boolean): Boolean {
-        return if (super.cancel(mayInterruptIfRunning)) {
-            deferred.cancel()
-            true
-        } else {
-            false
+    @Suppress("UnstableApiUsage")
+    override fun afterDone() {
```
Comment (yorickhenning) [inline integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: I vaguely remember there being a good reason not to use `afterDone()`. I'm pretty sure it came up last time around, maybe on an internal draft CR somewhere...

Seems wrong to override `cancel()` though. We should ask Chris.

Looks like I added test cases for `setFuture(cancelledFuture())`. Maybe there's some other completion-by-delegation corner case I've forgotten about.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 6]
```diff
@@ -421,42 +417,64 @@ private class OuterFuture<T>(private val deferred: Deferred<T>): ListenableFutur
...
 }
+
+/** Represents result of a `Coroutine` (either [Deferred] or [AbstractCoroutine]). */
+// This is similar to kotlin.Result type. However, kotlin.Result can't be used by external libraries.
+private sealed class Result<out T> {
+    class Success<T>(val value: T) : Result<T>()
+    class Failure<T>(val throwable: Throwable) : Result<T>()
+}
+
+/**
```
Comment (yorickhenning) [inline integration/kotlinx-coroutines-guava/src/ListenableFuture.kt:470]: These extension functions would probably be clearer as members of the type declaration, given that it's right here and all. :)

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt (hunk 4 above)
Comment (yorickhenning) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: Adding another patch or filing a bug are both reasonable plans here.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 7]
```diff
@@ -421,42 +417,64 @@ private class OuterFuture<T>(private val deferred: Deferred<T>): ListenableFutur
...
-) : AbstractFuture<Deferred<T>>() {
+private class InnerFuture<T>(
+    private val jobToCancel: Job
+) : AbstractFuture<Result<T>>() {
 
-    fun complete() {
-        set(deferred)
-    }
+    /**
+     * When entangled coroutine [completes][Job.isCompleted] (either successfully or exceptionally),
```
Comment (yorickhenning) [inline integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: As above with the parenthetical.

Related, Kotlin's Job is _extremely_ specific about terms: https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-job/

Perhaps phrasing this as "When the attached coroutine [\`isCompleted\`][Job.isCompleted]" would be more specific?

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 8]
```diff
@@ -70,10 +70,21 @@ public fun <T> CoroutineScope.future(
...
+ * When the returned `Deferred` is [cancelled][Deferred.cancel], it will try to propagate the
  * cancellation to `this` `ListenableFuture`. Propagation will succeed, barring a race with the
  * `ListenableFuture` completing normally. This is the only case in which the returned `Deferred`
  * will complete with a different outcome than `this` `ListenableFuture`.
+ *
+ * Note: Cancellation of the coroutine that calls [await][Deferred.await] on the returned
+ * deferred, **does not** cancel [Deferred] and original [ListenableFuture].
+ * To propagate cancellation from awaiting coroutine to the [ListenableFuture], use [ListenableFuture.await].
+ *
+ * ```
```
Comment (yorickhenning) [inline integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: Why is this? The cancellation should be propagated no matter the adaption.

I added [this](https://github.com/vadimsemenov/kotlinx.coroutines/blame/c4b4e1c0211818ed3d01bdf988e7749cbb95f910/integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt#L384) deferred-as-LF-as-deferred test case last time around. Did I miss a case, or is one of the `await()` implementations not cancellation-correct?

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 9]
```diff
@@ -70,10 +70,21 @@ public fun <T> CoroutineScope.future(
...
+ * suspend fun foo() {
+ *   listenableFuture.asDeferred().await()  // Doesn't propagate `foo`'s cancellation to the `listenableFuture`
```
Comment (yorickhenning) [inline integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: We're in the weeds on semantics, but `foo` isn't cancellable, because it's a function. The example isn't quite complete without a coroutine invoking `foo`. Using a `launch` coroutine builder might be easier to write technically correct comments on, when explaining calling-coroutine cancellation.

File: integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt [hunk 10]
```diff
@@ -540,12 +569,100 @@ class ListenableFutureTest : TestBase() {
         finish(3)
     }
 
+    @Test
+    fun testFutureScopeCancellationPropagatedAlong() = runTest {
```
Comment (yorickhenning) [inline integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt]: `testCancellingFutureContextJobCancelsFuture`

File: integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt [hunk 11]
```diff
@@ -540,12 +569,100 @@ class ListenableFutureTest : TestBase() {
...
+            foo.invokeOnCompletion {
+                expect(3)
+            }
+            val bar = async<Int> { throw TestException() }
+            foo.await() + bar.await()
+        }
+        yield()
+        expect(2)
+        // Blocking get should succeed after internal coroutine completes.
+        assertFailsWith<ExecutionException> { future.get() }.run {
```
Comment (yorickhenning) [inline integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt]: Use `getDone()` if the future is known to be complete.

File: integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt [hunk 12]
```diff
@@ -540,12 +569,100 @@ class ListenableFutureTest : TestBase() {
...
+            expect(4)
+            assertTrue(cause is TestException)
```
Comment (yorickhenning) [inline integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt]: This is a moderately exciting use of `run {}`. Reading this requires the reader to know that `assertFailsWith` returns the `Exception`, and to mentally track several layers of implicit typing.

This'd be less functionally ambitious as:

```
val thrown = assertFailsWith<ExecutionException> { getDone(future) }
assertTrue(thrown.cause is TestException)
```

File: integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt [hunk 13]
```diff
@@ -540,12 +569,100 @@ class ListenableFutureTest : TestBase() {
...
+                } finally {
+                    withContext(NonCancellable) {
+                        delay(200)
+                    }
+                }
+            }
+            foo.invokeOnCompletion {
+                expect(3)
+            }
+            val bar = async<Int> { throw TestException() }
```
Comment (yorickhenning) [inline integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt]: To strengthen the `is` below, create a `TestException` `val` scoped to the `runTest {}` block, throw that object instance here, and use object equality to assert that the expected exception was thrown, rather than a type introspection.

It's not likely to be some other `TestException` in this case, but it's the stronger statement, and it's *really* asserting that the expected exception was thrown. It's important with `IllegalStateException` in more complex cases.

File: integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt [hunk 14]
```diff
@@ -315,6 +315,27 @@ class ListenableFutureTest : TestBase() {
         finish(4)
     }
 
+    @Test
+    fun testAsListenableFuturePropagatesCancellationBeforeRunningListeners() = runTest {
```
Comment (yorickhenning) [inline integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt:321]: Interesting...

Does this test case pass before the production code changes?

**Review by yorickhenning — CHANGES_REQUESTED** [review]

File: integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt (hunk 14 above)
Comment (vadimsemenov) [reply integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt:321]: This one didn't. 
That's the problem of overriding `AbstractFuture.cancel` and calling `super.cancel` before cancelling the deferred. Listeners are executed inside `AbstractFuture.cancel`.

Comment (vadimsemenov) [reply integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt:321]: Similar testFuturePropagatesCancellationBeforeRunningListeners (for `CoroutineScope.future`) passed. But only because of AbstractFuture's implementation detail — it runs listeners in order of addition. And the listener that cancels the coroutine is the first one.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt (hunk 2 above)
Comment (vadimsemenov) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: There may be a race with external future cancellation.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt (hunk 8 above)
Comment (vadimsemenov) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: Cancellation of the Deferred will cancel original ListenableFuture, that's fine. However, cancellation of _the coroutine that calls `Deferred.await`_ doesn't cancel the Deferred.

This is kind of "by design" of the Deferred. Deferred represents a computation that can be `await`ed multiple times. So one can say that it's unfair to cancel the Deferred if a coroutine that `await`s on the Deferred is cancelled. Because other coroutines that `await` on the same Deferred may succeed. 

Here is a simple example: https://pl.kotl.in/DYcyjkku2
If you replace the `async` coroutine builder with `future`, future will be cancelled as expected. 
I think it's the main difference between `ListenableFuture.await` and `Deferred.await`. `ListenableFuture.await` is intended for one-shot future. `Deferred.await` assumes that the deferred can be reused.

You can also take a look at [awaitFutureWithCancel](https://github.com/Kotlin/kotlinx.coroutines/blob/master/integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt#L549) in tests.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt (hunk 4 above)
Comment (vadimsemenov) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: That's true. 
I'll try to normalise the exception at the moment of coroutine completion.
It means that stack trace of `ExecutionException` would be somewhat messed. But I hope no one looks at the stack trace of `ExecutionException`...

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt (hunk 3 above)
Comment (vadimsemenov) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: It was a surprise to me as well: https://stackoverflow.com/a/52632328

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt (hunk 2 above)
Comment (vadimsemenov) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: Reworded by referencing to Job.isCancelled.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt (hunk 9 above)
Comment (vadimsemenov) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: Fair enough. Replaced with `launch`.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt (hunk 7 above)
Comment (vadimsemenov) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: I agree. Reworded.

File: integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt (hunk 13 above)
Comment (vadimsemenov) [reply integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt]: Done.

File: integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt (hunk 11 above)
Comment (vadimsemenov) [reply integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt]: In this particular case it may not be done at this point. Future's coroutine is working in a different thread pool, so I indeed want to call a blocking get here. 
I didn't manage to find a better way to test that future isDone only after coroutine isCompleted. Can you see a better option?

File: integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt (hunk 12 above)
Comment (vadimsemenov) [reply integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt]: Done.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt (hunk 6 above)
Comment (vadimsemenov) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt:470]: Done.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt (hunk 4 above)
Comment (vadimsemenov) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: Done.

File: integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt (hunk 11 above)
Comment (yorickhenning) [reply integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt]: Ah, I misunderstood. `get()` is the right way to block the test thread awaiting completion, then.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt (hunk 8 above)
Comment (yorickhenning) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: OK, I'm following now.

I recommend striking the paragraph, then. The behaviour of `Deferred` isn't what this is documenting, it's documenting the adapter's behaviour.

Describing the cancellation of a calling coroutine and `await()` is adjacent to what this method is doing, and it confused me. I might need to care about that behaviour, but I should be discovering it through reading about `await()`, not here.

Some examples might be warranted, but I'm inclined to include positive examples rather than this one - cancellation works here and there, possibly contrasted with ordinary `Deferred` cancellation. It's also fair to leave those to the unit tests.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt (hunk 5 above)
Comment (vadimsemenov) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: After our offline discussion, reverted back to override `cancel()` and added a TODO to reconsider this when afterDone API is finalised.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt (hunk 8 above)
Comment (vadimsemenov) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: Removed the paragraph. I think `await()` documents its own implementation good enough.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 15]
```diff
@@ -407,49 +405,25 @@ private class OuterFuture<T>(jobToCancel: Job): ListenableFuture<T> {
...
-     *
-     * [OuterFuture] collaborates with this class to present a more cohesive picture and ensure
-     * that certain combinations of cancelled/cancelling states can't be observed.
      */
-    @Suppress("UnstableApiUsage")
-    override fun afterDone() {
-        if (isCancelled) {
+    override fun cancel(mayInterruptIfRunning: Boolean): Boolean {
+        // TODO: call jobToCancel.cancel() _before_ running the listeners.
+        //  Future.cancel may execute all the listeners which can delay cancellation of jobToCancel.
```
Comment (yorickhenning) [inline integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: "innerFuture.cancel() will execute innerFuture's listeners. This delays cancellation of jobToCancel until after innerFuture's listeners have already run." 

Something like that?

The distinction between "innerFuture" and "this Future" is a bit squishy. Maybe naming innerFuture "delegate" or something would be clearer. Not sure.

File: integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt [hunk 16]
```diff
@@ -316,6 +317,7 @@ class ListenableFutureTest : TestBase() {
     }
 
     @Test
+    @Ignore  // TODO: propagate cancellation before running listeners.
```
Comment (yorickhenning) [inline integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt:320]: File and link an issue in the TODO(). If not with this PR, in a followup.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 17]
```diff
@@ -195,12 +196,18 @@ private fun ExecutionException.nonNullCause(): Throwable {
...
  * This is inherently a race. See [Future.cancel] for a description of `Future` cancellation
  * semantics. See [Job] for a description of coroutine cancellation semantics. See
- * [DeferredListenableFuture.cancel] for greater detail on the overlapped cancellation semantics and
+ * [OuterFuture.cancel] for greater detail on the overlapped cancellation semantics and
  * corner cases of this method.
  */
 public fun <T> Deferred<T>.asListenableFuture(): ListenableFuture<T> {
     val outerFuture = OuterFuture<T>(this)
-    outerFuture.afterInit()
+    // This invokeOnCompletion completes the innerFuture as `deferred` does. The innerFuture may
```
Comment (yorickhenning) [inline integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: s/does/is completed/

Right?

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 18]
```diff
@@ -415,48 +405,54 @@ private class OuterFuture<T>(private val deferred: Deferred<T>): ListenableFutur
...
+            jobToCancel.cancel()
             true
         } else {
             false
         }
     }
 }
+
+/**
+ * Represents result of a `Coroutine` in the sense of [Future].
```
Comment (yorickhenning) [inline integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: s/result/the result/

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 19]
```diff
@@ -415,48 +405,54 @@ private class OuterFuture<T>(private val deferred: Deferred<T>): ListenableFutur
...
+    open val isCancelled get() = false
+
+    class Success<T>(val value: T) : Result<T>()
+    class Failure<T>(val exception: Exception) : Result<T>() {
+        override val isCancelled: Boolean get() = exception is CancellationException
+    }
+}
+
+/**
+ * Wraps exception in a `Coroutine` into [Result.Failure]. [Result.Failure.exception] represents `this`
```
Comment (yorickhenning) [inline integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: "Wraps an Exception thrown from a coroutine into" perhaps.

**Review by yorickhenning — APPROVED** [review]
Looks good! Probably time to send it upstream.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt (hunk 19 above)
Comment (vadimsemenov) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: Done. Emphasised that the exception is the receiver.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt (hunk 18 above)
Comment (vadimsemenov) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: Done.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt (hunk 17 above)
Comment (vadimsemenov) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: Reworded.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt (hunk 15 above)
Comment (vadimsemenov) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: Fixed the comment.

It's not exactly a "delegate", right? Renamed `innerFuture` to `auxFuture` and `OuterFuture` to `JobListenableFuture`. Let me know if you can think of a better naming.

File: integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt (hunk 16 above)
Comment (vadimsemenov) [reply integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt:320]: Yep, will do.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt (hunk 15 above)
Comment (yorickhenning) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: No, I don't have a better name to hand. "innerFuture", "actualFuture" or "internalFuture", anything like that...

The inner Future is the concrete implementation, and the outer Future is a proxy type that both owns the reference and invokes methods on it. I haven't read the Gang of Four in long enough to describe this design pattern better.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 20]
```diff
@@ -333,130 +327,133 @@ private class ListenableFutureCoroutine<T>(
...
         // reaching the "cancelled" state with a cause of CancellationException is treated as the
-        // same thing as innerFuture getting cancelled. If the Job is in the "cancelling" state and
+        // same thing as auxFuture getting cancelled. If the Job is in the "cancelling" state and
         // this Future hasn't itself been successfully cancelled, the Future will return
         // isCancelled() == false. This is the only discovered way to reconcile the two different
         // cancellation contracts.
-        return isDone
-          && (innerFuture.isCancelled
-          || deferred.getCompletionExceptionOrNull() is kotlinx.coroutines.CancellationException)
+        return isDone &&
```
Comment (charlesmunger) [inline integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: You can avoid the isDone check in some cases by doing it *after* you check isCancelled

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 21]
```diff
@@ -333,130 +327,133 @@ private class ListenableFutureCoroutine<T>(
...
+    }
+}
+
+/**
+ * Wraps `this` Exception thrown from a coroutine into [Result.Failure]. [Result.Failure.exception] represents `this`
+ * exception in the sense of [Future].
+ */
+private fun <T> Throwable.asResult() = Result.Failure<T>(
+    if (this is CancellationException) this
+    else ExecutionException(this)
```
Comment (charlesmunger) [inline integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: Both CancellationException and ExecutionException will have the wrong stack trace here. You can instead change result to be a sealed type with Success and Cancellation - use settablefuture.setException() for failures. 

This would give you settablefuture's ExecutionException wrapping for free; you'd create a new CancellationException in get(), with the input one as the cause.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 22]
```diff
@@ -333,130 +327,133 @@ private class ListenableFutureCoroutine<T>(
...
- *     could probably be compressed into one subclass of [AsyncFuture] to save an allocation, at the
+ *     [AbstractFuture.setFuture], is to just use an [AbstractFuture].
+ *   - To maintain sanity, this class implements [ListenableFuture] and uses an auxiliary [SettableFuture]
+ *     around coroutine's result as a state engine to establish happens-after-completion. This
+ *     could probably be compressed into one subclass of [AbstractFuture] to save an allocation, at the
  *     cost of the implementation's readability.
  */
-private class OuterFuture<T>(private val deferred: Deferred<T>): ListenableFuture<T> {
-    val innerFuture = DeferredListenableFuture(deferred)
+private class JobListenableFuture<T>(private val jobToCancel: Job): ListenableFuture<T> {
```
Comment (charlesmunger) [inline integration/kotlinx-coroutines-guava/src/ListenableFuture.kt:338]: Propagate toString() as done in AbstractFuture

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt (hunk 20 above)
Comment (vadimsemenov) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: Done.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt (hunk 21 above)
Comment (vadimsemenov) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: Thanks, done.
